contracts/token/erc1155/abstract/ERC1155Permit.Sol

```diff
@@ -0,0 +1,159 @@
+
+    bytes32 private immutable _PERMIT_TYPEHASH =
+        keccak256("Permit(address owner,address spender,bool approved,uint256 nonce,uint256 deadline)");
+
+    mapping(address => uint256) private _nonces;
+
+    constructor(string memory name, string memory uri)
+        ERC1155(uri)
+        EIP712(name, "1")
+    {}
+
+    /**
+     * @notice Overrides supportsInterface from IERC165 and ERC721Hybrid to add support for IERC4494.
+     * @param interfaceId The interface identifier, which is a 4-byte selector.
+     * @return True if the contract implements `interfaceId` and the call doesn't revert, otherwise false.
+     */
+    function supportsInterface(bytes4 interfaceId)
+      public
+      view
+      virtual
+      override(ERC1155)
+      returns (bool)
+    {
+     return
+      interfaceId == type(IERC1155Permit).interfaceId || // 0x5604e225
+      super.supportsInterface(interfaceId);
+    }
+
+    function permit(address owner, address spender, bool approved, uint256 deadline, bytes memory sig) external {
+        if (deadline < block.timestamp) {
+            revert PermitExpired();
+        }
+
+        bytes32 digest = _buildPermitDigest(spender, owner, approved, deadline);
+
+        // smart contract signature validation
+        if (_isValidERC1271Signature(owner, digest, sig)) {
+             _setApprovalForAll(owner, spender, approved);
+            return;
+        }
+
+        address recoveredSigner;
+
+        // EOA signature validation
+        if (sig.length == 64) {
+            // ERC2098 Sig
+            recoveredSigner = ECDSA.recover(
+                digest,
+                bytes32(BytesLib.slice(sig, 0, 32)),
+                bytes32(BytesLib.slice(sig, 32, 64))
+            );
+        } else if (sig.length == 65) {
+            // typical EDCSA Sig
+            recoveredSigner = ECDSA.recover(digest, sig);
+        } else {
+            revert InvalidSignature();
+        }
+
+        if (_isValidEOASignature(recoveredSigner, owner)) {
+            _setApprovalForAll(owner, spender, approved);
+        } else {
+            revert InvalidSignature();
+        }
+    }
+
+    /**
+     * @notice Builds the EIP-712 compliant digest for the permit.
+     * @param spender The address which is approved to spend the token.
+     * @param owner The address that holds the tokens.
+     * @param deadline The deadline until which the permit is valid.
+     * @return A bytes32 digest, EIP-712 compliant, that serves as a unique identifier for the permit.
+     */
+    function _buildPermitDigest(
+        address spender,
+        address owner,
+        bool approved,
+        uint256 deadline
+    ) internal returns (bytes32) {
+        return _hashTypedDataV4(
+            keccak256(
+                abi.encode(
+                    _PERMIT_TYPEHASH,
+                    owner,
+                    spender,
+                    approved,
+                    _nonces[owner]++,
```

Comment:
In the ERC721 hybrid we increment the nonce in an overriden `_transfer` function. Why is the implementation here different?

Reply:
The difference here is the number of tokens we give permissions to. I.e for erc721 its only one token, and since the `spender` is always validated, any replay attacks with the permit just approves the same spender to move that one token. 

Even though right now the permit gives an approveAll permission, if we wanted to add support for `approve(tokenid, amount)` in the future we would want this nonce to increment at permit validation to prevent replay attacks.